src/pop_manager.erl

```diff
@@ -0,0 +1,318 @@
+      } = PopManager,
+    
+
+    R1 = pop_chain:find_block_by_id(Hash, PC),
+    case R1 of
+	{ok, _} ->
+	    in_chain;
+	error ->
+	    R2 = maps:find(Hash, UB),
+	    case R2 of
+		{ok, _} ->
+		    unbound;
+		error ->
+		    unknown
+	    end
+    end.
+
+%% get previous block
+get_prev(Block, PC) ->
+    {ok, Prev} = pop_chain:find_block_by_id(maps:get(prev_id, Block), PC),
+    Prev.
+
+%% get n'th previous block, stop at genisys block
+get_nth_prev_block(N, Block, PC) ->
+    Ht = maps:get(height, Block),
+    if (Ht == 0) or (N == 0) ->
+	    Block;
+       (Ht > 0) and (N > 0) ->
+	    get_nth_prev_block(N-1, get_prev(Block, PC), PC)
+    end.
+
+setup_range_request(UnknownBlock, PopManager) ->
+    PC = PopManager#pop_manager.pop_chain,
+    N = PopManager#pop_manager.config#pop_manager_config.request_range_backup,
+
+    KnownBlock1 = pop_chain:get_head_block(PC),
+    KnownBlock2 = get_nth_prev_block(N, KnownBlock1, PC),
+    
+    {maps:get(this_id, KnownBlock1), maps:get(this_id, KnownBlock2), maps:get(this_id, UnknownBlock)}.
+
+%% given unknown block request, and longest known block from another party
+%% make a best guess about which blocks are unknown to the other party
+%% and return a array of hashes of those blocks
+
+compute_block_hash_range(UnknownBlockHash, KnownBlockHash, PopManager) ->
+    PC = PopManager#pop_manager.pop_chain,
+    UnknownBlock = pop_chain:find_block_by_id(UnknownBlockHash, PC),
+    KnownBlock = pop_chain:find_block_by_id(KnownBlockHash, PC),
+
+    compute_block_hash_range(UnknownBlock, KnownBlock, PC, []).
+
+compute_block_hash_range(UnknownBlock, KnownBlock, PC, AccList) ->
+    UnknownId = maps:get(this_id, UnknownBlock),
+    KnownId = maps:get(this_id, KnownBlock),
+
+    UnknownHeight = maps:get(height, UnknownBlock),
+    KnownHeight = maps:get(height, KnownBlock),
+
+    if UnknownId == KnownId ->
+	    AccList;  % done, found common parent
+
+       UnknownId /= KnownId ->
+	    %% decrement higher block, or both if equal height
+	    %% each time Unknown is decremented, add its hash to the list
+
+	    if UnknownHeight >= KnownHeight -> 	
+		    UnknownBlockNew = get_prev(UnknownBlock, PC), 
+	            KnownBlockNew = KnownBlock,
+	            AccListNew = [UnknownId | AccList];
+
+	       UnknownHeight =< KnownHeight ->
+		    UnknownBlockNew = UnknownBlock, 
+	            KnownBlockNew = get_prev(KnownBlock, PC),
+	            AccListNew = AccList;
+
+	       UnknownHeight == KnownHeight ->
+		    UnknownBlockNew = get_prev(UnknownBlock, PC), 
+	            KnownBlockNew = get_prev(KnownBlock, PC),
+	            AccListNew = [UnknownId | AccList]
+	    end,
+
+	    compute_block_hash_range(UnknownBlockNew, KnownBlockNew, PC, AccListNew)
+    end.
+    
+%% Adds a new block into pop_chain.
+%% Then it checks unbound blocks, and adds all that become bound.
+%% Function hook on_new_block is triggered on each addition.
+
+add_block_in_order(Block, CurrentTime, PopManager0) ->
```

Comment:
what does "in order" mean here?

Reply:
also what happens if you call this with a block that can't be added?

Reply:
"in order means" that the block is not a duplicate and can go right after an existing block

this function will fail on an assert otherwise

I only call it from add_block_out_of_order once I make sure it can be added